### ISSUE TITLE
keras.backend.eval takes increasing amounts of time with every call

### ISSUE THREAD

**Issue by sanderland** [keras-team/keras#12916]
**System information**  
- Have I written custom code (as opposed to using example directory):   yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):   Ubuntu 18.04
- TensorFlow backend (yes / no):   yes
- TensorFlow version:   1.13
- Keras version:   2.2.4
- Python version:   3.7.3
- CUDA/cuDNN version:   no CUDA

**Describe the current behavior**  
keras.backened.eval takes exponentially increasing time with every  call

**Describe the expected behavior**  
The same evaluation taking the same amount of time.

**Code to reproduce the issue**  
```
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
indim = 28
model = Sequential([Dense(8,input_shape=(indim,),activation='tanh'),Dense(4,activation='tanh'),Dense(1,activation='linear')])
model.compile(optimizer='adam',loss='mae')
def foo():
    for i in range(0,50):
        input = np.random.rand(32,28)
        Y     = np.random.rand(32,1)
        Ypred = model.predict(input)
        loss = model.loss_functions[0](Y,Ypred)
        loss = keras.backend.eval(loss)
```
### this outputs 1 seconds, 2 seconds, 4 seconds as results for subsequent runs
```
%timeit foo()
```
### profiler output below
```
%prun foo()
```

**Other info / logs**  
Confirmed/replicated on another PC.

Profile output shows ExtendSession and TF_SessionRun_wrapper taking all the time.  
```
        166454 function calls (162854 primitive calls) in 2.156 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       50    1.098    0.022    1.098    0.022 {built-in method _pywrap_tensorflow_internal.ExtendSession}
       50    0.896    0.018    0.896    0.018 {built-in method _pywrap_tensorflow_internal.TF_SessionRun_wrapper}
      200    0.015    0.000    0.016    0.000 tf_stack.py:31(extract_stack)
       50    0.011    0.000    0.011    0.000 {built-in method _pywrap_tensorflow_internal.TF_SessionRunCallable}
      100    0.005    0.000    0.076    0.001 op_def_library.py:339(_apply_op_helper)
      200    0.003    0.000    0.003    0.000 {built-in method _pywrap_tensorflow_internal.TF_FinishOperation}
      200    0.003    0.000    0.086    0.000 deprecation.py:473(new_func)
      200    0.003    0.000    0.049    0.000 ops.py:1688(__init__)
```

Maybe related to Issue #6620 but since that has no resolution and is years old.

**Comment by Dref360** [comment]
You are creating new ops every time you call the loss.

This wouldn't happen if you were using model.evaluate or tf.eager_mode.
Here is your example with eager_mode activated:

```
import numpy as np
import time
import tensorflow as tf
from tensorflow import keras

from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
import tensorflow as tf

tf.enable_eager_execution()
indim = 28
model = Sequential([Dense(8, input_shape=(indim,), activation='tanh'), Dense(4, activation='tanh'),
                    Dense(1, activation='linear')])
model.compile(optimizer=tf.train.AdamOptimizer(), loss='mae')

for i in range(0, 50):
    s = time.time()
    input = np.random.rand(32, 28)
    Y = np.random.rand(32, 1)
    Ypred = model.predict(input)
    loss = model.loss_functions[0](Y, Ypred).numpy()
    print("Took", time.time() - s)
```

**Comment by jvishnuvardhan** [comment]
@sanderland Can you check the suggestion by @Dref360 and let us know if this is resolved or not. thanks!

**Comment by sanderland** [comment]
@Dref360 Your solution works in isolation, but since the code is going into a module, I have no control over whether tf.enable_eager_execution() was called at program startup or which model modules are used.
model.evaluate does not return the predicted output, and I want to pass both the loss and predicted value on, without running the model twice (as predict+evaluate would require)

And sure I'm creating all sorts of new objects every iteration (new input, Y, Ypred, graph, etc) -- but why should eval take exponentially increasing amounts of time? 

It does seem that loss.graph is a single global which contains an increasing number of items, and nothing is ever removed.
It's also not 'eval', but calls like Y = keras.backend.abs(Y) which add to this list. Wrapping the loss function call and evaluation in another 'with tf.Session() as sess:' doesn't help either.

**Comment by Dref360** [comment]
Without eager_mode, you need to make your own placeholder.

Something like:
```
import time

import numpy as np

import keras.backend as K
from keras.layers import Dense
from keras.models import Sequential

indim = 28
model = Sequential([Dense(8, input_shape=(indim,), activation='tanh'), Dense(4, activation='tanh'),
                    Dense(1, activation='linear')])
model.compile(optimizer='adam', loss='mae')
y_pred = K.placeholder([None, 1])
y_true = K.placeholder([None, 1])
loss_fn = model.loss_functions[0](y_true, y_pred)

for i in range(0, 500):
    s = time.time()
    input = np.random.rand(32, 28)
    Y = np.random.rand(32, 1)
    Ypred = model.predict(input)
    _ = K.get_session().run(loss_fn, feed_dict={y_true: Y, y_pred: Ypred})
    print("Took", time.time() - s)

```

**Comment by sanderland** [comment]
Alright, thanks @Dref360
